### DISCUSSION TITLE
Llama object has no attribute 'ctx'

### DISCUSSION THREAD

**Discussion by HorrySheet** [oobabooga/text-generation-webui#1615] [Q&A]
Um.. so I thought I followed the instructions and I cant seem to get this thing to run any models I stick in the folder and have it download via hugging face.. always gives something around the lines of this error .. What did I do wrong? Something in the installation process or what?

 Gradio HTTP request redirected to localhost :)
bin C:\Oobabooga\installer_files\env\lib\site-packages\bitsandbytes\libbitsandbytes_cpu.dll
C:\Oobabooga\installer_files\env\lib\site-packages\bitsandbytes\cextension.py:33: UserWarning: The installed version of bitsandbytes was compiled without GPU support. 8-bit optimizers, 8-bit multiplication, and GPU quantization are unavailable.
  warn("The installed version of bitsandbytes was compiled without GPU support. "
Loading xzuyn_pygmalion-6B-v3-ggml-q4_3...
llama.cpp weights detected: models\xzuyn_pygmalion-6B-v3-ggml-q4_3\ggml-model-q4_3.bin

Traceback (most recent call last):
  File "C:\Oobabooga\text-generation-webui\server.py", line 914, in <module>
    shared.model, shared.tokenizer = load_model(shared.model_name)
  File "C:\Oobabooga\text-generation-webui\modules\models.py", line 141, in load_model
    model, tokenizer = LlamaCppModel.from_pretrained(model_file)
  File "C:\Oobabooga\text-generation-webui\modules\llamacpp_model_alternative.py", line 30, in from_pretrained
    self.model = Llama(**params)
  File "C:\Oobabooga\installer_files\env\lib\site-packages\llama_cpp\llama.py", line 111, in __init__
    self.ctx = llama_cpp.llama_init_from_file(
  File "C:\Oobabooga\installer_files\env\lib\site-packages\llama_cpp\llama_cpp.py", line 156, in llama_init_from_file
    return _lib.llama_init_from_file(path_model, params)
OSError: [WinError -1073741795] Windows Error 0xc000001d
Exception ignored in: <function Llama.__del__ at 0x0000016462BEBD00>
Traceback (most recent call last):
  File "C:\Oobabooga\installer_files\env\lib\site-packages\llama_cpp\llama.py", line 804, in __del__
    if self.ctx is not None:
AttributeError: 'Llama' object has no attribute 'ctx'

Done!
Press any key to continue . . .

**Comment by electric-samurai** [comment]
I had a similar error message but not the source of the issue doesn't seem the same.  In my case I had specified the wrong path to my model.  However, it led me to investigate the code right before the error, it is attempting to load the model into self.ctx.  Maybe there is something wrong with your bin file so it can't load it?  Have to tried to load that model with llama.cpp, or tried other models?

**Comment by HorrySheet** [comment]
I just discovered llama.ccp or whatever  And I'm trying to wrap my head
around what the difference between that and kibold and tavern really are
..  It seems like this technology has evolved so fast and there are some
ways that have come out to make it easier but some people are still using
some mixed matched methods for whatever reason or another. I'm so confused.
I think my issue with this error is because of permissions on this stupid
@ss work computer and the fact that I cannot run things in administrator
mode. So it's probably trying to do some read or write stuff and doesn't
have proper permissions. Either way I give up trying to run this locally..


 What I'm really trying to do is run the best role playing chat bot
simulator that can utilize text to speech functions like Silero. So I
figure the Pygmalion model is what I should try to run..  So far I'm having
a hell of a time getting the character to remember anything about itself
compared to running that same character with vicuna..

Can anyone suggest a YouTube Earth that is talking primarily about role
playing and these chat models?  I really want to make some cool game like a
mud or something.

**Comment by markusantonuis** [comment]
@HorrySheet  I'm getting a similar error to the above. Did you manage to find a solution?
  File "D:\Python311\Lib\site-packages\pyllamacpp\model.py", line 137, in __del__
    if self._ctx:
       ^^^^^^^^^
AttributeError: 'GPT4All' object has no attribute '_ctx'

**Reply by chenhongze1** [reply]
do you deal with it?

**Comment by ManuelZ** [comment]
I'm on Windows and had a similar-looking problem:
```
    if self.ctx is not None:
AttributeError: 'Llama' object has no attribute 'ctx'
```

 Instead of passing the path like this:
```
model_path="F:\LLMs\alpaca_7B\ggml-model-q4_0.bin"
```

I passed it like this:
```
model_path="F:\\LLMs\\alpaca_7B\\ggml-model-q4_0.bin"
```

**Comment by HorrySheet** [comment]
Thanks but it seems there is a whole other issue going in with it. I'm
taking a break for now.

**Comment by rrggrr** [comment]
Provide an absolute path to your models directory:

`/./Users/<user>/<path to models>`

**Comment by markusantonuis** [comment]
I tried all kinds of configurations of the file path but still getting the same error:
model = GPT4All("C:\\Users\\BF\\AppData\\Local\\nomicai\\GPT4All\\ggml-gpt4all-l13b-snoozy.bin")
model = GPT4All("C:/Users/BF/AppData/Local/nomicai/GPT4All/ggml-gpt4all-l13b-snoozy.bin")

Note the first line has two backslashes at each folder level but they don't show in the comment here

**Comment by hootan-alavizadeh** [comment]
To fix this you need to modify this line in your source code

# for vicuna model 7b
CTX_MAX = 8192

llm = Llama(
    model_path=MODEL_PATH,
    n_ctx=CTX_MAX, n_threads=THREADS_NUM,
    use_mlock=True,
    verbose=False,
)

**Reply by TFWol** [reply]
Which file are you referring to?

**Comment by minfuel** [comment]
**This worked for us:**

**Issue** #0
**Description:** file loading issues
**Fix:** 
- 	pip install llama-cpp-python==0.1.48 resolved my issue, along with
- 	pip install pygpt4all==v1.0.1 --force-reinstall

**when using**
- 	https://huggingface.co/mrgaang/aira/blob/main/gpt4all-converted.bin
- 	https://huggingface.co/Pi3141/alpaca-native-7B-ggml/blob/main/ggml-model-q4_0.bin

_we also changed the name from 'ggml-model-q4_0.bin' to 'ggml_model_q4_0.bin'
(basically removing the dashes w/ underscore)_

**Reply by freakskyy** [reply]
YOUR FKN AWESOME MAN!! KING

**Reply by TFWol** [reply]
Have you gotten GGLM models from [The Bloke](https://huggingface.co/TheBloke) working?

I don't want to use the models you point to since I find it odd you upvoted your own post and another account was recently created who did the same...

**Reply by minfuel** [reply]
Indeed, we employ the cutting-edge GGLM models. The models I mentioned are merely for reference purposes and should not be utilized. Furthermore, they are strictly for personal use, and I neither possess nor have any affiliation with them. These models represent the solution I discovered to rectify the issues I encountered with the ones I initially employed. I upvoted my own post as I firmly believe it will alleviate the headaches experienced by many individuals. As to my knowledge this solution should work on any GGML model

**Reply by TFWol** [reply]
Understood.
I had tried setting `llama-cpp-python` to other versions without being able to get around the ctx error.

I've yet to pull repo changes from the past couple of weeks to see if the problem is still around.

**Reply by TFWol** [reply]
[Found the problem](https://github.com/LostRuins/koboldcpp/issues/317#issue-1803287493). I'll have to build for noAVX2 combination

**Reply by minfuel** [reply]
Great work, hopefuly it's better than my downgrading option... I think i will try this myself

**Reply by LightTemplar** [reply]
I had 0.1.57 version and this problem. I've just upgraded to 0.1.77 and the problem is gone.

**Reply by TFWol** [reply]
@LightTemplar 
Just curious, does your CPU have AVX2 support?

Easiest way I know to check on windows:
Using Sysinternals [coreinfo](https://learn.microsoft.com/en-us/sysinternals/downloads/coreinfo)
```batch
coreinfo | findstr /i avx2
```

My result (`*` is Yes, `-` is no):
`AVX2            -       Supports AVX2 instruction extensions`
Mine is `-`, so I don't

Alternatively you could use **gcc**:
```shell
gcc -v -xc /dev/null -O3 -march=native -o- -E 2>&1 | grep -o -- '-m\(no-\)\?avx2'
```
If you see `-mno-avx2` then AVX2 isn't supported on the CPU

**Reply by TFWol** [reply]
Never mind, solution for text-gen is [this comment](https://github.com/oobabooga/text-generation-webui/issues/3276#issuecomment-1648532571)

**Comment by belalesawe1** [comment]
You have to reinstall llama-cpp-python using the command below to fix it:

pip install llama-cpp-python --force-reinstall --upgrade --no-cache-dir
